aten/src/ATen/CPUGenerator.cpp

```diff
@@ -158,13 +152,33 @@ void CPUGenerator::set_next_double_normal_sample(c10::optional<double> randn) {
   next_double_normal_sample_ = randn;
 }
 
+/**
+ * Get the engine of the CPUGenerator
+ */
+at::mt19937 CPUGenerator::engine() {
+  return engine_;
+}
+
+/**
+ * Set the engine of the CPUGenerator
+ * 
+ * See Note [Thread-safety and Generators]
+ */
+void CPUGenerator::set_engine(at::mt19937 engine) {
+  engine_ = engine;
+}
+
 /**
  * Private clone method implementation
  * 
  * See Note [Thread-safety and Generators]
  */
 CloneableGenerator<CPUGenerator, Generator>* CPUGenerator::clone_impl() const {
-  return new CPUGenerator(engine_);
+  auto gen = new CPUGenerator();
+  gen->set_engine(engine_);
+  gen->set_next_float_normal_sample(next_float_normal_sample_);
+  gen->set_next_double_normal_sample(next_double_normal_sample_);
+  return gen;
```

Comment:
I missed this on my first review, but could we have this return a `std::unique_ptr<CloneableGenerator<CPUGenerator, Generator>>` instead of a raw pointer that is manually memory managed? If you want to do this in a follow-up PR, that's OK  (if this is the only problem I'll initiate land without a fix.)

Reply:
This pointer gets wrapped by a unique pointer: https://github.com/pytorch/pytorch/pull/16604/files#diff-621abddec4e628a1cd833e601c92751eR14 . It has to remain a raw pointer to support covariant unique_ptr return type: https://www.fluentcpp.com/2017/09/12/how-to-return-a-smart-pointer-and-use-covariance/

Reply:
Oh, I forgot about that. Disregard!